### PR TITLE
Add typecheck for variant type

### PR DESCRIPTION
**Pull request by swamper123**
Fix for #451 and #450 discussion.

Tested with our own Client, since UaExpert won't allow such a request for normal Nodes.

### CODE REVIEW

**Comment by swamper123** [comment]
It seems that some tests probem appear with that PR.
Either it's my fault and it breaks, or the tests are not well enough (and I will fix them).

**Comment by swamper123** [comment]
Just a question:
```python
async def test_service_fault(server, admin_client):
    request = ua.ReadRequest()
    request.TypeId = ua.FourByteNodeId(999)  # bad type!
    with pytest.raises(ua.UaStatusCodeError):
        await admin_client.uaclient.protocol.send_request(request)
```
This test fails, because UaStatusCodeError BadTypeMismatch raised....but it expects any UaStatusCode Error, does it?

**Comment by swamper123** [comment]
It seems that:

https://github.com/FreeOpcUa/opcua-asyncio/blob/80a4cc2635e8888ddda051f6dc282e2873570d0e/tests/util_enum_struct.py#L29

is making trouble.
Inside of Node 6009, the value attribute is empty. Type is VariantType.Null, Value is None etc., but the attribute DataType seems to be right.

grafik

I will not change write_attribute_value, since this could be used for any attribut. Not sure where stuff is causing trouble here atm.

**Comment by oroulet** [comment]
Sorry I was very late looking at that PR. 
I think the principle is OK, if there if VariantType is not the expected one we should raise type mismatch (apart from Vraianttype.Null that can replace most of the other types).
It is going to break some things, but we probably need to fix these things ,also all people writting Int64 where UInt32 is expected will break too..

But why do you exclude all these standard nodes?

File: asyncua/server/address_space.py [hunk 1]
```diff
@@ -706,6 +706,41 @@ async def write_attribute_value(self, nodeid, attr, value):
         if attval is None:
             return ua.StatusCode(ua.StatusCodes.BadAttributeIdInvalid)
 
+        if value.Value.VariantType != attval.value.Value.VariantType:
+            if not (nodeid.NamespaceIndex == 0 and nodeid.Identifier in (
```
Comment (oroulet) [inline asyncua/server/address_space.py:710]: not (xxx and ddd in (ccc)) is far too complicated test. I am sure you can change the order to make it understadnable.

some like 

if xxxx:
   pass
do_somehing()

**Comment by oroulet** [comment]
I made an alterntive where all tests pass just by correctly handling NULL type

**Comment by oroulet** [comment]
https://github.com/FreeOpcUa/opcua-asyncio/pull/587

**Comment by swamper123** [comment]
> Sorry I was very late looking at that PR.
No worries, haven't touched it in a while anyway. 😅 

> I think the principle is OK, if there if VariantType is not the expected one we should raise type mismatch (apart from Vraianttype.Null that can replace most of the other types).

A VariantType.Null that will be overwritten by something or something that will be overwritten by VariantType.Null?

> But why do you exclude all these standard nodes?

I am not sure anymore tbh...I think something broke in the tests with these notes when they are initialized. Changing the Type was not expected in this draft. I will study the specs again, what the excepted behaviour is.

Allowing exceptions for this case would may still lead to problems. Otherwise each half-broken client could change int32 to Null and the Problem would still exist.

**Comment by oroulet** [comment]
>> Allowing exceptions for this case would may still lead to problems. Otherwise each half-broken client could change int32 to Null and the Problem would still exist.

Valid point. Does spec says anything about that? maybe only admin should be allowed to write Null or maybe nobody...

I close that one. comment on the other PR

**Pull request closed without merging**